### PR TITLE
Dynamic blacklist and refactor

### PR DESCRIPTION
**Pull request by neuroscr**
address #9 which required a bit of planning and refactoring

- added database/models (right now memory driver only)
- a global moderator can now blacklist a pubkey
  - will invalidate all their tokens on platform
- moves "globals" (moderators) out of the loki.ini into database
- start database for sessions (tempdb) but not yet hooked up
- also includes user/mod mass delete (adapted to new format)

# new structure:
want to reduce the number of lines per file and make sure we have a plan as we grew.

dialects/ 
There's currently 2 dialects: token and moderation
The file server needs token but not moderation
has a group of 3 files: routes, handlers, helpers

models/
database models and functions

logic/
general internal api function that dialects can use

### CODE REVIEW

File: logic/permissions.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  },
+  addGlobalModerator: async userid => {
+    const result = await storage.addServerModerator(userid);
+  },
+  blacklistUserFromServer: async userid => {
+    // have to promisify these to play nicely with other promise stuff
+    // actually maybe I didn't need these...
+
+    // take an id return a user object
+    function getUserPromise(userid) {
```
Comment (BeaudanBrown) [inline logic/permissions.js]:
```suggestion
    const getUserPromise = (userid) => {
```

Comment (BeaudanBrown) [reply logic/permissions.js]: Etc below

File: models/role_permissions.js [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        if (perm_obj[perm] === -1) perms[perm] = -1;
+        // if 1 enable
+        else if (perm_obj[perm] === 1) perms[perm] = 1;
+      })
+    })
+    if (perms['moderator'] !== -1 && perms['moderator'] !== 1) {
+      // FIXME: look up server default
+      // write back to perms['allowed']
+    }
+    return perms['moderator'] == 1 ? true : false;
```
Comment (BeaudanBrown) [inline models/role_permissions.js]: Redundant ternary?
Also is `==` intended?

File: models/roles.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+module.exports = {
+  start: start,
+  getRoleIdByName: (name) => {
+    return new Promise((resolve, rej) => {
+      roleModel.find({ where: { name } }, function(err, roles) {
+        if (err) {
+          console.error(err)
+          return rej(err)
+        }
+        if (roles.length != 1) {
```
Comment (BeaudanBrown) [inline models/roles.js]:
```suggestion
        if (roles.length !== 1) {
```

File: models/roles.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+          return cb('length ' + roles.length, false)
+        }
+        //cb(err, roles[0].id)
+        resolve(roles[0].id)
```
Comment (BeaudanBrown) [inline models/roles.js]: Are we sure that roles will always be of length >== 1 here?
Will get undefined error otherwise

File: storage.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
+const funcs = [];
+funcs.push(require('./models/users.js'));
+funcs.push(require('./models/challenges.js'));
+funcs.push(require('./models/roles.js'));
+funcs.push(require('./models/user_roles.js'));
+funcs.push(require('./models/role_permissions.js'));
+
+const Schema = require('caminte').Schema
+
+memoryUpdate = function (model, filter, data, callback) {
```
Comment (BeaudanBrown) [inline storage.js]: Scoping without const here intended?
```suggestion
memoryUpdate = (model, filter, data, callback) => {
```

File: storage.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    this.all(model, filter, function(err, nodes) {
+      var count = nodes.length;
+      if (!count) {
+        return callback(false, cleanData);
+      }
+      nodes.forEach(function(node) {
+        mem.cache[model][node.id] = Object.assign(node, cleanData);
+        if (--count === 0) {
+          callback(false, cleanData);
+        }
```
Comment (BeaudanBrown) [inline storage.js]: Not sure I understand here, looks like this is just meant to trigger after the forEach has executed on every element?
In which case can just put the callback call after the forEach and remove the count?

File: storage.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    //schema.automigrate(function() {});
+    // don't lose data
+    schema.autoupdate(function() {});
+  }
+}
+
+let functions = {};
+funcs.forEach((func) => {
+  functions = Object.assign(functions, func);
+});
```
Comment (BeaudanBrown) [inline storage.js]: I believe Object.assign mutates, so enough to do:
```js
const functions = {};
funcs.map(func => Object.assign(functions, func));
```
Also not reassigning functions so can use const

File: storage.js (hunk 5 above)
Comment (neuroscr) [reply storage.js]: should have a const

File: dialects/moderation/dialect_moderation_handlers.js [hunk 8]
```diff
@@ -0,0 +1,256 @@
+const helpers = require('./dialect_moderation_helpers');
+
+// all input / output filtering should happen here
+
+let cache, dialect
+const setup = (utilties) => {
+  // config are also available here
+  ({ cache, dialect, logic, storage } = utilties);
```
[the remark this answers is not in the document]
Comment (neuroscr) [reply dialects/moderation/dialect_moderation_handlers.js]: should be scoped in that let

File: dialects/moderation/dialect_moderation_helpers.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+      const allowed = access_list.indexOf(message.channel_id);
+      if (allowed === -1) {
+        // not allowed to manage this channel
+        const resObj={
+          meta: {
+            code: 403,
+            error_message: "You're not allowed to moderation this channel"
+          }
+        };
+        return resObj;
```
[the remark this answers is not in the document]
Comment (neuroscr) [reply dialects/moderation/dialect_moderation_helpers.js]: it's meant to be pushed to metas/deltas

File: models/roles.js (hunk 4 above)
Comment (neuroscr) [reply models/roles.js]: yup, the previous if makes sure of that, length will always exactly one here.

File: logic/permissions.js [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    function getUserByUsernamePromise(pubKey) {
+      return new Promise(function(resolve, rej) {
+        cache.getUserID(pubKey, (user, err) => {
+          if (err) console.error('logic:::permissions::blacklistUserFromServer - getUserID err', err);
+          resolve(user);
+        });
+      });
+    }
+
+    const user = await getUserByUsernamePromise(pubKey);
```
[the remark this answers is not in the document]
Comment (neuroscr) [reply logic/permissions.js]: I think other functions will eventually call it and will be refactored out eventually...

**Pull request merged**